### DISCUSSION TITLE
A Problem about Phase Missing

### DISCUSSION THREAD

**Discussion by MWW726** [pycalphad/pycalphad#390] [Q&A] [answered]
![screenshot](https://user-images.githubusercontent.com/85090681/148058695-09ab4c96-8ae2-4ba0-a635-129b1bc3e2ab.jpg)
I want to know How to solve the problem of phase missing in low or some middle temperature.

**Comment by bocklund** [comment, marked as answer]
For low temperatures, pycalphad will not automatically extrapolate outside of the Piecewise temperature limits like you may be used to in other software. That is, if the TDB defines

```
FUNCTION GHSERAL  298.15 <expression>; 2000 N !
```

the Gibbs energy would be strictly defined between 298.15 K and 2000 K and the Gibbs energy will be zero otherwise. Please double check that all the temperatures are within the temperature limits of the TDB.

**Reply by MWW726** [reply]
Thank you. Got it. But, I still can`t understand phase missing in 500K while no phase missing in 450K and 550K. In fact, I found phase suddenly  miss sometimes in Piecewise temperature limits. do you have some suggestions?
![image](https://user-images.githubusercontent.com/85090681/148180755-1c7b7b4f-9120-49b1-b533-920396993d29.png)

**Reply by bocklund** [reply]
Can you share the database file? I can try to reproduce the issue and find a solution.

**Reply by MWW726** [reply]
[al-cu.txt](https://github.com/pycalphad/pycalphad/files/7819766/al-cu.txt)

**Reply by bocklund** [reply]
A small update here. I am able to reproduce the issue. At this composition, `ALCU_ETA` becomes the only stable phase and there's a discontinuity in the chemical potentials at this point, which causes pycalphad to get stuck in a loop of trying to add `ALCU_THETA`, `ALCU_ZETA`, `ALCU_DELTA` phase, depending on the chemical potentials/driving forces.

![image](https://user-images.githubusercontent.com/7681751/150271831-6027e190-52d7-43bf-9f1f-15653a740941.png)

![image](https://user-images.githubusercontent.com/7681751/150272801-6c775f5c-90e5-44e0-a3a8-9cb788340429.png)

This is a bug in the minimizer that may have been introduced when we changed some solver internals. Depending on your application, a possible workaround is to nudge the composition to be slightly Cu-rich or Cu-poor, e.g.

```python
eq = equilibrium(dbf, comps, phases, {v.P: 101325, v.T: 450, v.X("CU"): 0.50 + 1e-6, v.N: 1}, output='HM', verbose=False)
```

This will give you two phases, but the phase fraction of the second phase will be small.

**Reply by Wuwei-Mo** [reply]
Hello, recently, I have been performing equilibrium calculations for aluminum alloys with multiple components. I have encountered the same issue of missing phases that we discussed before. However, this time, I have more elements, and the values are constrained to 5-6 decimal places, yet the problem persists. Additionally, since the TDB file for this calculation belongs to our research group, I cannot share it. Is there a solution to this issue?
![1](https://github.com/pycalphad/pycalphad/assets/129748194/c64100eb-24f7-4fc2-8c08-be5857ad8c95)
![2](https://github.com/pycalphad/pycalphad/assets/129748194/18d0739d-f5b7-494e-82e2-084ad33f09a7)
